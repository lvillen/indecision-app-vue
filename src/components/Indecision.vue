<template>
  <img src="https://via.placeholder.com/250" alt="bg">
  <div class="bg-dark"></div>

  <div class="indecision-container">
    <input
      v-model="question"
      type="text"
      placeholder="Ask a question!"
    >    
    <p>Remember to end with a question mark: "?"</p>

    <div class="question">
      <h2>{{ question }}</h2>
      <h1>No way, punk</h1>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      question: null
    }
  },
  watch: {
    question(value, oldValue) {
      if(!value.includes('?')) return 

      //TODO http request
    }
  }
}
</script>

<style>
    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }
</style>