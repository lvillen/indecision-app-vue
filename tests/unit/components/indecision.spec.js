import { shallowMount } from "@vue/test-utils";
import Indecision from '@/components/Indecision';

describe('Indecision Component', () => {
  let wrapper;
  let consoleLogSpy;

  global.fetch = jest.fn( () => Promise.resolve({
    json: () => Promise.resolve({
        'answer': 'yes',
        'forced': false,
        'image': 'https://yesno.wtf/assets/yes/2.gif'
      })
    }) 
  ); //Cleaning the error of node not having fetch and mocking our call to the YesNo API

  beforeEach(() => {
    wrapper = shallowMount(Indecision)
    consoleLogSpy = jest.spyOn( console, 'log');

    jest.clearAllMocks(); //Clear the data from the mocks used in each test.
  });

  test('Should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Should not call to getAnswer function when writing (console.log)', async() => {
    const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer' ); //TODO: Refactor
    
    const input = wrapper.find('input');
    await input.setValue('Hello world');

    expect(consoleLogSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).not.toHaveBeenCalled();
  });
  
  test('Should call to getAnswer function when writing a question mark "?"', async() => {
    const getAnswerSpy = jest.spyOn( wrapper.vm, 'getAnswer'); //TODO: Refactor

    const input = wrapper.find('input');
    await input.setValue('Have getAnswer been called?');

    expect(consoleLogSpy).toHaveBeenCalledTimes(1);
    expect(getAnswerSpy).toHaveBeenCalled();
  });

  test('Testing getAnswer calling fetch', async() => {
    await wrapper.vm.getAnswer();

    const img = wrapper.find('img');

    expect(img.exists()).toBeTruthy();
    expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif');
    expect(wrapper.vm.answer).toBe('yes');

  });

  test('Testing getAnswer â€” API fails', async() => {
    fetch.mockImplementationOnce( () => Promise.reject('API down'));

    await wrapper.vm.getAnswer();

    const img = wrapper.find('img');

    expect(img.exists()).toBeFalsy();
    expect(wrapper.vm.answer).toBe('Could not load from the API');
  });
})