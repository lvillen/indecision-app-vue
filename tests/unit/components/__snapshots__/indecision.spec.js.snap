// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Should match the snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Ask a question!">
  <p>Remember to end with a question mark: "?"</p>
  <!--v-if-->
</div>
`;
